### PR TITLE
Add the various MLT io classes - Used for serializing collected stacktraces

### PR DESCRIPTION
**Pull request by tylerbenson**
Binary format described in https://docs.google.com/document/d/1GQFRVWi8oVKWAk2XvoEGW1_PceSKTuJld8BaVOnshaE

Co-authored-by: Jean-Philippe Bempel <jean-philippe.bempel@datadoghq.com>
Co-authored-by: Jaroslav Bachorik <jaroslav.bachorik@datadoghq.com>
Co-authored-by: Laplie Anderson <laplie.anderson@datadoghq.com>

### CODE REVIEW

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/LEB128Writer.java [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+  LEB128Writer writeChar(char data);
+
+  /**
+   * Write {@linkplain Character} data in LEB128 encoding at the given offset
+   *
+   * @param offset the offset from which to start writing the data
+   * @param data the data
+   * @return the writer position after the data has been written
+   */
+  long writeChar(long offset, char data);
```
Comment (lpriima) [inline utils/mlt-support/src/main/java/com/datadog/mlt/io/LEB128Writer.java:37]: Java uses `int` as index of array and we can't create array more than `Integer.MAX_VALUE` . 
I don't understand why we are using `long` as offset in this interface and then we have to convert it to `int` anyway in  many places in implementation:
https://github.com/DataDog/dd-trace-java/pull/1708/files#diff-d12c79eb0010a77cc4ca68b1b31d4225R43

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTChunkCollector.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+import java.util.stream.Stream;
+import lombok.Getter;
+
+public abstract class MLTChunkCollector implements IMLTChunk {
+  /*
+   * TODO seems like subtree compression is worse than plain full-tree deduplication
+   *  when the subtree compression is finally removed this flag should go as well + subtree support in FrameSequence
+   */
+  private static final boolean USE_SUBTREE_COMPRESSION =
+      Boolean.getBoolean("mlt.subtree_compression");
```
Comment (lpriima) [inline utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTChunkCollector.java:18]: Do we need dd. prefix and/or ENV var support, Config and etc ... for this one ?

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTReader.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
+package com.datadog.mlt.io;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/** The MLT binary format reader */
+public final class MLTReader {
```
Comment (lpriima) [inline utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTReader.java:8]: class without state? Should we mark all methods as static and add private constructor ?

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
+import it.unimi.dsi.fastutil.ints.IntSet;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.function.Consumer;
+import java.util.function.IntConsumer;
+
+/** The MLT binary format writer */
+public final class MLTWriter {
```
Comment (lpriima) [inline utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java:11]: class without state? Should we mark all methods as static and add private constructor ?

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/LEB128Writer.java (hunk 1 above)
Comment (jbachorik) [reply utils/mlt-support/src/main/java/com/datadog/mlt/io/LEB128Writer.java:37]: This started with using mmapped file as the backing storage in mind. Hence 'long' index. But I agree that for our current (and any foreseeable future) usage an 'int' offset would be sufficient.

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTChunkCollector.java (hunk 2 above)
Comment (jbachorik) [reply utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTChunkCollector.java:18]: This should be gone - the subtree compression proved to be suboptimal for the data we are collecting for MLT. The flag as well as all the related functionality can safely be removed.

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java (hunk 4 above)
Comment (jbachorik) [reply utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java:11]: The eternal question whether to have all-static singletons or instance singletons :)
I have no strong opinion either way - only that mocking would be impossible for all-static one.

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTReader.java (hunk 3 above)
Comment (bantonsson) [reply utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTReader.java:8]: 👍

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java (hunk 4 above)
Comment (bantonsson) [reply utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java:11]: 👍

Comment (lpriima) [reply utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java:11]: you don't have to mock something which doesn't have state :)

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/ConstantPool.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
+package com.datadog.mlt.io;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+
+/**
+ * A generic constant pool implementation.
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply utils/mlt-support/src/main/java/com/datadog/mlt/io/ConstantPool.java]: Done

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java (hunk 4 above)
Comment (jbachorik) [reply utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTWriter.java:12]: Ok, all-static now In theory, you still can mock no-state class - eg. a factory impl which will return mock instanes in turn. But it is not the case here.

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTChunkCollector.java (hunk 2 above)
Comment (jbachorik) [reply utils/mlt-support/src/main/java/com/datadog/mlt/io/MLTChunkCollector.java]: Subtree compression removed. No need for the config any more.

File: utils/mlt-support/src/main/java/com/datadog/mlt/io/LEB128Writer.java (hunk 1 above)
Comment (jbachorik) [reply utils/mlt-support/src/main/java/com/datadog/mlt/io/LEB128Writer.java]: Refactored to `int` offset values.

**Comment by jbachorik** [comment]
Added some rudimentary tests for `MLTReader` and `MLTWriter`.
Also addressed (I think) all comments/requests.

**Pull request merged**